File: compiler/output/servo-grammar.js
//@type action
function inlineflow_height () { return 0; }
//@type action
function blockflow_box__position ( _ale_arg0) { return _ale_arg0; }
//@type action
function blockflow_my_height ( _ale_arg0) { return getHeight(_ale_arg0); }
//@type action
function blockflow_height ( _ale_arg0,  _ale_arg1) { return (_ale_arg0 == Au(0)) ? _ale_arg1 : _ale_arg0; }
function isInorder(node, pass) {
  switch (pass) {
    case 0:
     throw "did not expect inorder call for pass 0";
    case 1:
     throw "did not expect inorder call for pass 1";
    case 2:
     throw "did not expect inorder call for pass 2";
    default: throw ("unknown pass " + pass);
  }
}
function visit_blockflow_0(node) {
  logger.log("  visit  BlockFlow (id: " + node.id + "): " + 0);
  setAttribSafe(node, "my_height", blockflow_my_height(getAttribSafe(getChildByRefName(node,"box_"), "style")));
  logger.log("    blockflow_my_height: " + getAttribSafe(node, "my_height"));
  logger.log("        box_@style: " + getAttribSafe(getChildByRefName(node,"box_"), "style"));
  return true;
}
function visit_blockflow_1(node) {
  logger.log("  visit  BlockFlow (id: " + node.id + "): " + 1);

  setAttribSafe(node, "flow_children_bottom_init", (1));
  setAttribSafe(node, "flow_children_bottom_last", getAttribSafe(node, "flow_children_bottom_init"));
  logger.log("      init flow_children@bottom: " + getAttribSafe(node, "flow_children_bottom_init"));
  logger.log("    last init flow_children_bottom_last: " + getAttribSafe(node, "flow_children_bottom_last"));
  (function () {
    var children = getChildren(node, "flow_children", false);
    for (var i = 0; i < children.length; i++) {
      var child = children[i]; 
      setAttribSafe(child, "bottom", (getAttribSafe(i == 0 ? node : children[i-1], i == 0 ? ("flow_children_bottom_init") : ("bottom")) + getAttribSafe(child, "height") ));
      logger.log("         step flow_children@bottom: " + getAttribSafe(child, "bottom"));
    }
  })();


  setAttribSafe(node, "flow_children_position_init", (0));
  setAttribSafe(node, "flow_children_position_last", getAttribSafe(node, "flow_children_position_init"));
  logger.log("      init flow_children@position: " + getAttribSafe(node, "flow_children_position_init"));
  logger.log("    last init flow_children_position_last: " + getAttribSafe(node, "flow_children_position_last"));
    setAttribSafe(node, "childs_height_init", (0));
  setAttribSafe(node, "childs_height", getAttribSafe(node, "childs_height_init"));
  logger.log("      init childs_height: " + getAttribSafe(node, "childs_height_init"));
  logger.log("    last init childs_height: " + getAttribSafe(node, "childs_height"));
  (function () {
    var children = getChildren(node, "flow_children", false);
    for (var i = 0; i < children.length; i++) {
      var child = children[i]; 
      setAttribSafe(child, "position", (Rect(Au(0), getAttribSafe(child, "bottom") - getAttribSafe(child, "height"), Au(100), getAttribSafe(child, "height"))));
      logger.log("         step flow_children@position: " + getAttribSafe(child, "position"));
      setAttribSafe(node, "childs_height", (getAttribSafe(node, "childs_height") + rectHeight(getAttribSafe(child, "position"))));
      logger.log("         step childs_height: " + getAttribSafe(node, "childs_height"));
    }
  })();

  setAttribSafe(node, "height", blockflow_height(getAttribSafe(node, "my_height"), getAttribSafe(node, "childs_height")));
  logger.log("    blockflow_height: " + getAttribSafe(node, "height"));
  logger.log("        my_height: " + getAttribSafe(node, "my_height"));
  logger.log("        childs_height: " + getAttribSafe(node, "childs_height"));
  return true;
}
function visit_blockflow_2(node) {
  logger.log("  visit  BlockFlow (id: " + node.id + "): " + 2);
  setAttribSafe(getChildByRefName(node,"box_"), "position", blockflow_box__position(getAttribSafe(node, "position")));
  logger.log("    blockflow_box__position: " + getAttribSafe(getChildByRefName(node,"box_"), "position"));
  logger.log("        position: " + getAttribSafe(node, "position"));
  return true;
}
function visit_genericbox_0(node) {
  logger.log("  visit  GenericBox (id: " + node.id + "): " + 0);
  return true;
}
function visit_genericbox_1(node) {
  logger.log("  visit  GenericBox (id: " + node.id + "): " + 1);
  return true;
}
function visit_genericbox_2(node) {
  logger.log("  visit  GenericBox (id: " + node.id + "): " + 2);
  return true;
}
function visit_inlineflow_0(node) {
  logger.log("  visit  InlineFlow (id: " + node.id + "): " + 0);
  setAttribSafe(node, "height", inlineflow_height());
  logger.log("    inlineflow_height: " + getAttribSafe(node, "height"));
  return true;
}
function visit_inlineflow_1(node) {
  logger.log("  visit  InlineFlow (id: " + node.id + "): " + 1);
  return true;
}
function visit_inlineflow_2(node) {
  logger.log("  visit  InlineFlow (id: " + node.id + "): " + 2);
  return true;
}
function visit_0 (node, isGlobalCall, parent) {
  if (node.nodeType == 1) {
    switch (node.tagName.toLowerCase()) {
      case "blockflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_blockflow_0(node);
      case "inlineflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_inlineflow_0(node);
      case "genericbox":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_genericbox_0(node);
    }
  }
  if (node.nodeType == 3) { logger.log("skipping text node 2"); return; }
  throw ("unknown node type for dispatch: " + node.tagName);
}
function visit_1 (node, isGlobalCall, parent) {
  if (node.nodeType == 1) {
    switch (node.tagName.toLowerCase()) {
      case "blockflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_blockflow_1(node);
      case "inlineflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_inlineflow_1(node);
      case "genericbox":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_genericbox_1(node);
    }
  }
  if (node.nodeType == 3) { logger.log("skipping text node 2"); return; }
  throw ("unknown node type for dispatch: " + node.tagName);
}
function visit_2 (node, isGlobalCall, parent) {
  if (node.nodeType == 1) {
    switch (node.tagName.toLowerCase()) {
      case "blockflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_blockflow_2(node);
      case "inlineflow":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_inlineflow_2(node);
      case "genericbox":
        logger.log("global: " + isGlobalCall + ", parent: " + parent);
        return visit_genericbox_2(node);
    }
  }
  if (node.nodeType == 3) { logger.log("skipping text node 2"); return; }
  throw ("unknown node type for dispatch: " + node.tagName);
}
function layout (root) {
  inherit(visit_0, root);
  synthesize(visit_1, root);
  inherit(visit_2, root);
}

File: compiler/output/servo-grammar.rs
extern mod libftl;
use libftl::{base,Node,log,MidNode,Leaf,inherit,synthesize};
use std::int::to_str_bytes;
use std::util;
//@type action
fn inlineflow_height () -> Au { 0 }
//@type action
fn blockflow_box_ ( _ale_arg0: Au) -> Au { boxVal(_ale_arg0) }
//@type action
fn blockflow_my_height ( _ale_arg0: Au) -> Au { getHeight(_ale_arg0) }
//@type action
fn blockflow_height ( _ale_arg0: Au,  _ale_arg1: Au) -> Au { if ((_ale_arg0 == Au(0))) {_ale_arg1} else {_ale_arg0} }
impl FtlNode for InlineFlow{
 fn with_all_children(&mut self, func: |&mut FtlNode|) {
 }
 fn visit_0(&mut self) {
  log("  visit  InlineFlow (id: " + self.id.to_str() + "): " + "0");
  self.base.height = (inlineflow_height());
  log("    inlineflow_height: " + self.base.height.to_str());
 
 }
 fn visit_1(&mut self) {
  log("  visit  InlineFlow (id: " + self.id.to_str() + "): " + "1");
 
 }
 fn visit_2(&mut self) {
  log("  visit  InlineFlow (id: " + self.id.to_str() + "): " + "2");
 
 }
}

impl FtlNode for BlockFlow{
 fn with_all_children(&mut self, func: |&mut FtlNode|) {
  for child in self.flow_children.mut_iter() {
    func(*child);
  }
 }
 fn visit_0(&mut self) {
  log("  visit  BlockFlow (id: " + self.id.to_str() + "): " + "0");
  self.my_height = (blockflow_my_height(self.style));
  log("    blockflow_my_height: " + self.my_height.to_str());
  log("        style: " + self.style.to_str());
 
 }
 fn visit_1(&mut self) {
  log("  visit  BlockFlow (id: " + self.id.to_str() + "): " + "1");

  let flow_children_bottom_init = ((1));
  let mut flow_children_bottom_last = (flow_children_bottom_init);
  log("      init flow_children@bottom: " + flow_children_bottom_init.to_str());
  log("    last init flow_children_bottom_last: " + self.base.flow_children_bottom_last.to_str());
  let mut children = util::replace(&mut self.flow_children, ~[]);
  for child in children.mut_iter() {
      let child: &mut BaseFlow  = base(as_ftl_node(child));
      child.bottom = ((if first { flow_children_bottom_init } else { flow_children_bottom_last } + child.height ));
      flow_children_bottom_last = child.bottom;
      log("         step flow_children@bottom: " + child.bottom.to_str());
  }
  self.flow_children = children;

  let flow_children_position_init = ((0));
  let mut flow_children_position_last = (flow_children_position_init);
  log("      init flow_children@position: " + flow_children_position_init.to_str());
  log("    last init flow_children_position_last: " + self.base.flow_children_position_last.to_str());
    let childs_height_init = ((0));
  self.childs_height = (childs_height_init);
  log("      init childs_height: " + childs_height_init.to_str());
  log("    last init childs_height: " + self.childs_height.to_str());
  let mut children = util::replace(&mut self.flow_children, ~[]);
  for child in children.mut_iter() {
      let child: &mut BaseFlow  = base(as_ftl_node(child));
      child.position = ((Rect(Au(0), child.bottom - child.height, Au(100), child.height)));
      flow_children_position_last = child.position;
      log("         step flow_children@position: " + child.position.to_str());
      self.childs_height = ((self.base.childs_height + rectHeight(child.position)));
      log("         step childs_height: " + self.childs_height.to_str());
  }
  self.flow_children = children;
  self.base.height = (blockflow_height(self.my_height, self.childs_height));
  log("    blockflow_height: " + self.base.height.to_str());
  log("        my_height: " + self.my_height.to_str());
  log("        childs_height: " + self.childs_height.to_str());
 
 }
 fn visit_2(&mut self) {
  log("  visit  BlockFlow (id: " + self.id.to_str() + "): " + "2");
  self.box_ = (blockflow_box_(self.base.position));
  log("    blockflow_box_: " + self.box_.to_str());
  log("        position: " + self.base.position.to_str());
 
 }
}

fn main() {
    layout(&mut generateTree());
}
fn layout (root: &mut FtlNode) {
  inherit(|node| node.visit_0(), root);
  synthesize(|node| node.visit_1(), root);
  inherit(|node| node.visit_2(), root);
}
